src/cman.jl

```diff
@@ -59,268 +92,388 @@ type MPIManager <: ClusterManager
+                while nb_available(w_s) > 0
+                    data = takebuf_array(w_s.buffer)
+                    push!(reqs, MPI.Isend(data, rank, 0, mgr.comm))
+                end
+                if !isempty(reqs)
+                    (indices, stats) = MPI.Testsome!(reqs)
+                    filter!(req -> req != MPI.REQUEST_NULL, reqs)
                 end
+                # TODO: Need a better way to integrate with libuv's event loop
+                yield()
             end
+            put!(mgr.sending_done, nothing)
         end
-        nothing
+        (r_s, w_s)
+    catch e
+        Base.show_backtrace(STDOUT, catch_backtrace())
+        println(e)
+        rethrow(e)
     end
 end
 
-procs(manager::MPIManager) = sort([p for p in keys(manager.j2mpi)])
-mpiprocs(manager::MPIManager) = sort([p for p in keys(manager.mpi2j)])
-
+################################################################################
+# Alternative startup model: All Julia processes are started via an external
+# mpirun, and the user does not call addprocs.
 
-function start(mode=TCP_TRANSPORT_ALL)
-    comm, comm_size, rank = init_mpi()
+# Enter the MPI cluster manager's main loop (does not return on the workers)
+function start_main_loop(mode::TransportMode=TCP_TRANSPORT_ALL;
+                         comm::MPI.Comm=MPI.COMM_WORLD)
+    !MPI.Initialized() && MPI.Init()
+    @assert MPI.Initialized() && !MPI.Finalized()
     if mode == TCP_TRANSPORT_ALL
+        # Base is handling the workers and their event loop
+        # The workers have no manager object where to store the communicator.
+        # TODO: Use a global variable?
+        comm = MPI.COMM_WORLD
+        rank = MPI.Comm_rank(comm)
+        size = MPI.Comm_size(comm)
         if rank == 0
-            cman = MPIManager(;np=comm_size-1, mode=TCP_TRANSPORT_ALL)
-            for j in 1:(comm_size-1)
-                MPI.send((getipaddr().host, cman.port), j, j, comm)
+            # On the manager: Perform the usual steps
+            # Create manager object
+            mgr = MPIManager(np=size-1, mode=mode)
+            mgr.comm = comm
+            # Send connection information to all workers
+            # TODO: Use Bcast
+            for j in 1:size-1
+                MPI.send((getipaddr().host, mgr.port), j, 0, comm)
             end
-
-            addprocs(cman)
-            return cman
-         else
-            (obj, status) = MPI.recv(0, rank, comm)
+            # Tell Base about the workers
+            addprocs(mgr)
+            return mgr
+        else
+            # On a worker: Receive connection information
+            (obj, status) = MPI.recv(0, 0, comm)
             (host, port) = obj
+            # Call the regular worker entry point
             setup_worker(host, port) # does not return
-         end
-     elseif mode == MPI_TRANSPORT_ALL
+        end
+    elseif mode == MPI_TRANSPORT_ALL
+        #TODO comm = MPI.Comm_dup(comm)
```

Comment:
Oops, this code should be enabled. That is, it should be working in the case `MPI_TRANSPORT_ALL`.